### PR TITLE
drivers: usb: stm32: fix support of STM32U5 OTG_HS with embedded PHY

### LINKED ISSUE #79823: Add support to STM32 USB OTG-HS to re-define what is passed into HAL_SYSCFG_SetOTGPHYReferenceClockSelection()
**Is your enhancement proposal related to a problem? Please describe.**

The current Zephyr STM32 USB OTG-HS driver hard-codes a 16MHz clock source into the HSE clock.

From drivers/usb/device/usb_dc_stm32.c:

        HAL_SYSCFG_SetOTGPHYReferenceClockSelection(SYSCFG_OTG_HS_PHY_CLK_SELECT_1);

This means if you're using a different clock (in my case a 24MHz crystal), it needs a code change.

In my case I'm using an STMU5 part, the definitions are in the STM32 HAL:

stm32cube/stm32u5xx/drivers/include/stm32u5xx_hal.h 

Update Oct 16 - I've now verified that changing it to SELECT_4 works fine for my 24MHz crystal on the board.

**Describe the solution you'd like**

It would be nice to either show up as a device-tree config, or a Kconfig option. I'm going to prototype both.
 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mniestroj**
Introduce new binding `st,stm32u5-otghs-phy` for OTG_HS PHY. This allows to
configure clock source and handle STM32U5 specific OTG_HS PHY behavior in
driver implementation in a more readable way.

Move OTG_HS PHY clock selection (previously `<&rcc STM32_SRC_HSI48
ICKLK_SEL(0)>`) from OTG_HS node to OTG_HS PHY node.

Rename `USBPHYC_SEL` -> `OTGHS_SEL` which matches the definition in the stm32u5
`CCIPR2` register (RM0456 Rev 5, Section 11.8.47).

Support enabling OTG_HS PHY clock, which is bit 15 (`OTGHSPHYEN`) in
`RCC_AHB2ENR1`. Change OTG_HS clock to be bit 14 (`OTGEN`).

Calculate in runtime OTG_HS PHY clock source frequency. Try to match that
to supported (16, 19.2, 20, 24, 26, 32 MHz) frequencies and select proper
option with `HAL_SYSCFG_SetOTGPHYReferenceClockSelection()` API (instead of
hardcoded 16 MHz selection).

This is alternative implementation (but uses parts of it) to: #80249. Most important differences:
 * OTG_HS PHY clock source frequency is calculated in runtime, so no need to specify that manually
   in devicetree.
   - better developer experience, due to less options to be selected and aligned with each other in
     devicetree
   - automatically checks in runtime if selected frequency is supported

Fixes: #79823

Tested with `nucleo_u5a5zj_q` board with `zephyr/samples/subsys/usb/cdc_acm` sample, in following
scenarios:

 * default (`HSE` as 16MHz clock source): no devicetree modifications
 * `PLL1_P` as 16MHz clock source:

 ```
&pll1 {
    div-p = <20>;
};

&otghs_phy {
    clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
             <&rcc STM32_SRC_PLL1_P OTGHS_SEL(1)>;
};
 ```

 * `PLL1_P` as 32MHz clock source:

 ```
&pll1 {
    div-p = <10>;
};

&otghs_phy {
    clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
             <&rcc STM32_SRC_PLL1_P OTGHS_SEL(1)>;
};
 ```

 * `PLL1_P` as 20MHz clock source:

```
&pll1 {
    div-p = <16>;
};

&otghs_phy {
    clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
             <&rcc STM32_SRC_PLL1_P OTGHS_SEL(1)>;
};
```

 * `PLL1_P/2` as 20MHz clock source:

```
&pll1 {
    div-p = <8>;
};

&otghs_phy {
    clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
             <&rcc (STM32_SRC_PLL1_P | STM32_CLOCK_DIV(2)) OTGHS_SEL(3)>;
};
```

Depends on (and contains): #82637

### CODE REVIEW

File: dts/bindings/phy/st,stm32u5-otghs-phy.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
+                 <&rcc STM32_SRC_HSE OTGHS_SEL(2)>;
+
+        /* PLL1_P_CK */
+        clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
+                 <&rcc STM32_SRC_PLL1_P OTGHS_SEL(1)>;
+
+        /* PLL1_P_CK/2 */
+        clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
+                 <&rcc STM32_SRC_PLL1_P OTGHS_SEL(3)>;
```
Comment (erwango) [inline dts/bindings/phy/st,stm32u5-otghs-phy.yaml:36]: I would actually prefer we introduce a new `STM32_SRC_PLL1_P_DIV_2` source to avoid breaking `clock_control_get_rate()` api behavior (with your proposal it may not return the right rate anymore).

To make it reusable, we can think of adding a trick on the u32 value such as :
```
                 <&rcc (STM32_SRC_PLL1_P | DIV_2) OTGHS_SEL(3)>;
```
With the DIV_2 being extracted in `clock_control_get_rate()` and included in rate computation before returning (to make it functional with similar HSE/2, HSI48/2, HSI/256, MSIK/1024 and MSIK/4)

**Review by erwango — CHANGES_REQUESTED** [review]
Initial feedback:

I'm all in for the main idea of instantiating a distinct phy clock, thanks. I'm fine with the dynamic computation as well.
Though, one comment on the clocks binding "abuse".

File: dts/bindings/phy/st,stm32u5-otghs-phy.yaml (hunk 1 above)
Comment (mniestroj) [reply dts/bindings/phy/st,stm32u5-otghs-phy.yaml]: Great feedback!

I've implemented something along those lines, which allows to use:

```
<&rcc (STM32_SRC_PLL1_P | STM32_CLOCK_DIV(2)) OTGHS_SEL(3)>;
```

**Comment by FRASTM** [comment]
Fixing the build error  on west build -p auto -b nucleo_u5a5zj_q samples/boards/st/mco  with CI, requires 
```patch
diff --git a/include/zephyr/dt-bindings/clock/stm32u5_clock.h b/include/zephyr/dt-bindings/clock/stm32u5_clock.h
index 4b9b563121e..be159575518 100644
--- a/include/zephyr/dt-bindings/clock/stm32u5_clock.h
+++ b/include/zephyr/dt-bindings/clock/stm32u5_clock.h
@@ -139,4 +139,11 @@
 #define MCO1_SEL(val)           STM32_MCO_CFGR(val, 0xF, 24, CFGR1_REG)
 #define MCO1_PRE(val)           STM32_MCO_CFGR(val, 0x7, 28, CFGR1_REG)
 
+/* MCO prescaler : division factor */
+#define MCO_PRE_DIV_1 0
+#define MCO_PRE_DIV_2 4
+#define MCO_PRE_DIV_3 5
+#define MCO_PRE_DIV_4 6
+#define MCO_PRE_DIV_5 7
+
 #endif /* ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_STM32U5_CLOCK_H_ */
```

File: drivers/usb/device/usb_dc_stm32.c [hunk 2]
```diff
@@ -333,7 +422,7 @@ static int usb_dc_stm32_clock_disable(void)
 		LOG_ERR("Unable to disable USB clock");
 		return -EIO;
 	}
-#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) && defined(CONFIG_SOC_SERIES_STM32U5X)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy)
 	LL_AHB2_GRP1_DisableClock(LL_AHB2_GRP1_PERIPH_USBPHY);
 #endif
```
Comment (erwango) [inline drivers/usb/device/usb_dc_stm32.c:427]: Isn't this a `clock_control_off(clk, (clock_control_subsys_t)&phy_pclken[0]);` now?

**Review by erwango — DISMISSED** [review]
Thanks @mniestroj for the quick and complete update.

Only required change would be to update rcc bindings description with the new STM32_CLOCK_DIV() trick ?

Otherwiwse looks globaly good to me.

File: drivers/usb/device/usb_dc_stm32.c (hunk 2 above)
Comment (mniestroj) [reply drivers/usb/device/usb_dc_stm32.c:427]: Yes, it seems so. Should I replace it with `clock_control_off()` API?

Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c:427]: it would be nice.

Comment (mniestroj) [reply drivers/usb/device/usb_dc_stm32.c]: Done!

**Review by marwaiehm-st — APPROVED** [review]
LGTM

**Comment by erwango** [comment]
I'd like @erikarn feedback, but  otherwise LGTM.

**Review by FRASTM — APPROVED** [review]

**Comment by erwango** [comment]
One more question, maybe for @marwaiehm-st: 
How does it fit with usb next support ? 
At the minimum, we should be able to bring this back (or forward) to usb next driver. Does it sound doable ?
(Ideally, we should not break STM32U5 OTG_HS support on USB next, but it is possible it doesn't work well so far anyway)

**Comment by marwaiehm-st** [comment]
@erwango 
Yes, it is feasible. I do not foresee a significant issues with integrating the new st,stm32u5-otghs-phy into the USB next driver.

**Comment by erikarn** [comment]
lgtm, thanks for digging into the weeds on cleaning this stuff up!

**Comment by erikarn** [comment]
(I'm sad I won't get a commit into zephyr this time, but I'm sure there'll be others to do. I'll give this a run myself when it hits the mainline tree and provide some further testing feedback!)

**Comment by mniestroj** [comment]
> (I'm sad I won't get a commit into zephyr this time, but I'm sure there'll be others to do. I'll give this a run myself when it hits the mainline tree and provide some further testing feedback!)

One commit in this PR is yours, while I have acknowledged your input (with Co-authored-by) in the second one. This PR would not exist in this shape without your work #80249. So thanks a lot for that!

The reason I created this PR is because I spent some time testing and already got ideas on how to improve that and ask community for feedback. Overall I just wanted to push this forward as far as I could, as otherwise I would might not find time for it in the future :)

So thanks a lot @erikarn for your contribution!

**Comment by erikarn** [comment]
> > (I'm sad I won't get a commit into zephyr this time, but I'm sure there'll be others to do. I'll give this a run myself when it hits the mainline tree and provide some further testing feedback!)
> 
> One commit in this PR is yours, while I have acknowledged your input (with Co-authored-by) in the second one. This PR would not exist in this shape without your work #80249. So thanks a lot for that!
> 
> The reason I created this PR is because I spent some time testing and already got ideas on how to improve that and ask community for feedback. Overall I just wanted to push this forward as far as I could, as otherwise I would might not find time for it in the future :)
> 
> So thanks a lot @erikarn for your contribution!

ship itttttttt i have a bunch more usb stuff on my needing to do list :-) :-)

**Review by erwango — APPROVED** [review]

File: drivers/usb/device/usb_dc_stm32.c [hunk 3]
```diff
@@ -208,16 +208,67 @@ void HAL_PCD_SOFCallback(PCD_HandleTypeDef *hpcd)
 }
 #endif
 
-static int usb_dc_stm32_clock_enable(void)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy)
+
+static const struct stm32_pclken phy_pclken[] = STM32_DT_CLOCKS(DT_INST_PHANDLE(0, phys));
+
+static int usb_dc_stm32u5_phy_clock_select(const struct device *const clk)
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_stm32.c:215]: @erwango Why is it not in the clock control driver, and why do we have to deal with all this unrelated driver code here? How will the changes be synchronized with the new driver?

**Review by jfischer-no — CHANGES_REQUESTED** [review]
In the last commit, remove `(Issue https://github.com/zephyrproject-rtos/zephyr/issues/79823)` from the commit title.

File: drivers/usb/device/usb_dc_stm32.c (hunk 3 above)
Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c:215]: I understand your concern, but this configuration is purely out of clock control handling and depends on USB.
Note that even isn't a "RCC" block configuration, but a "SYSCFG" (System configuration controller) one.
Clock controller driver doesn't have information about the USB context, it "just" enables/disables clocks and has the knowledge about the related rate. Configuring which clock should be used it out if its scope and depend on consuming IP.

> How will the changes be synchronized with the new driver?
For sure, we're planning to port this to the new driver (see my above comment), but I didn't want to make this blocking for @mniestroj.

Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c:215]: @jfischer-no Please comment.

Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c:215]: FYI, just created https://github.com/zephyrproject-rtos/zephyr/issues/82745 for tracking

Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c:215]: @jfischer-no PTAL. Your legitimate concern about backport is I think addressed. For the other point about clock-control architecture, while I disagree, can be discussed later on but let's avoid scope creep in current PR.

**Review by jfischer-no — APPROVED** [review]

File: drivers/usb/device/usb_dc_stm32.c (hunk 3 above)
Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c:215]: > We also have drivers/usb/common, you can put code there that is used by the both drivers

Interesting thanks. We'll surely consider it. Adding to comment to `drivers/usb/common`
